File: go-server/internal/tools/feed.go
// Package tools provides MCP tool implementations
package tools

import (
	"context"
	"fmt"
	"strings"

	"github.com/oyin-bo/autoreply/go-server/internal/auth"
	"github.com/oyin-bo/autoreply/go-server/internal/bluesky"
	"github.com/oyin-bo/autoreply/go-server/internal/mcp"
	"github.com/oyin-bo/autoreply/go-server/pkg/errors"
)

// FeedTool implements the feed tool
type FeedTool struct {
	apiClient *bluesky.APIClient
	credStore *auth.CredentialStore
}

// NewFeedTool creates a new feed tool
func NewFeedTool() (*FeedTool, error) {
	apiClient, err := bluesky.NewAPIClient()
	if err != nil {
		return nil, fmt.Errorf("failed to create API client: %w", err)
	}

	credStore, err := auth.NewCredentialStore()
	if err != nil {
		return nil, fmt.Errorf("failed to initialize credential store: %w", err)
	}

	return &FeedTool{
		apiClient: apiClient,
		credStore: credStore,
	}, nil
}

// Name returns the tool name
func (t *FeedTool) Name() string {
	return "feed"
}

// Description returns the tool description
func (t *FeedTool) Description() string {
	return "Get the latest feed from BlueSky. Returns a list of posts (also known as tweets or skeets). If you want to see the latest posts from a specific user, provide their handle."
}

// InputSchema returns the JSON schema for tool input
func (t *FeedTool) InputSchema() mcp.InputSchema {
	return mcp.InputSchema{
		Type: "object",
		Properties: map[string]mcp.PropertySchema{
			"feed": {
				Type:        "string",
				Description: "(Optional) The feed to retrieve, can be a BlueSky feed URI, or a name for a feed to search for. If unspecified, returns the default popular feed 'What is Hot'.",
			},
			"viewAs": {
				Type:        "string",
				Description: "(Optional) Account to view feed as: handle (alice.bsky.social), @handle, DID, Bsky.app profile URL, or partial DID suffix. If unspecified or 'anonymous', the feed will be retrieved in incognito mode.",
			},
			"continueAtCursor": {
				Type:        "string",
				Description: "(Optional) Cursor for pagination. When provided, fetches the next batch of posts from where the previous request left off.",
			},
			"limit": {
				Type:        "integer",
				Description: "(Optional) Defaults to 50",
			},
		},
		Required: []string{},
	}
}

// Call executes the feed tool
func (t *FeedTool) Call(ctx context.Context, args map[string]interface{}, _ *mcp.Server) (*mcp.ToolResult, error) {
	// Extract parameters
	feed := getStringParam(args, "feed", "")
	viewAs := getStringParam(args, "viewAs", "")
	continueAtCursor := getStringParam(args, "continueAtCursor", "")
	limit := getIntParam(args, "limit", 50)

	// Validate and normalize viewAs
	if viewAs != "" && viewAs != "anonymous" {
		// Check if credentials exist
		_, err := t.credStore.Load(viewAs)
		if err != nil {
			// Try to get default handle
			defaultHandle, defErr := t.credStore.GetDefault()
			if defErr == nil && defaultHandle != "" {
				viewAs = defaultHandle
			} else {
				viewAs = "anonymous" // Fall back to anonymous
			}
		}
	}

	// If viewAs is empty, try to get default handle
	if viewAs == "" {
		defaultHandle, err := t.credStore.GetDefault()
		if err == nil && defaultHandle != "" {
			viewAs = defaultHandle
		}
	}

	// Fetch the feed - if limit > 100, fetch in batches
	var allPosts []interface{}
	var cursor string
	if continueAtCursor != "" {
		cursor = continueAtCursor
	}

	requestedLimit := limit
	if requestedLimit <= 0 {
		requestedLimit = 50
	}

	// Fetch in batches if needed
	for len(allPosts) < requestedLimit {
		batchSize := requestedLimit - len(allPosts)
		if batchSize > 100 {
			batchSize = 100 // API limit per request
		}

		params := make(map[string]string)
		if cursor != "" {
			params["cursor"] = cursor
		}
		params["limit"] = fmt.Sprintf("%d", batchSize)

		var feedData map[string]interface{}
		var err error

		if feed != "" {
			// Resolve feed URI if needed
			var feedURI string
			feedURI, err = t.resolveFeedURI(ctx, feed)
			if err != nil {
				return nil, errors.Wrap(err, errors.InvalidInput, "Failed to resolve feed")
			}

			// Use resolved feed URI
			params["feed"] = feedURI
			feedData, err = t.apiClient.GetWithOptionalAuth(ctx, viewAs, "app.bsky.feed.getFeed", params)
		} else if viewAs != "" && viewAs != "anonymous" {
			// Get authenticated user's timeline
			feedData, err = t.apiClient.GetWithAuth(ctx, viewAs, "app.bsky.feed.getTimeline", params)
		} else {
			// Use default "What's Hot" feed for anonymous users
			params["feed"] = "at://did:plc:z72i7hdynmk6r22z27h6tvur/app.bsky.feed.generator/whats-hot"
			feedData, err = t.apiClient.GetPublic(ctx, "app.bsky.feed.getFeed", params)
		}

		if err != nil {
			return nil, errors.Wrap(err, errors.InternalError, "Failed to fetch feed")
		}

		// Extract posts from this batch
		batchPosts, ok := feedData["feed"].([]interface{})
		if !ok || len(batchPosts) == 0 {
			// No more posts available
			break
		}

		allPosts = append(allPosts, batchPosts...)

		// Check if there's a cursor for next batch
		nextCursor, hasCursor := feedData["cursor"].(string)
		if !hasCursor || nextCursor == "" {
			// No more pages available
			break
		}
		cursor = nextCursor

		// If we got fewer posts than requested in this batch, we've reached the end
		if len(batchPosts) < batchSize {
			break
		}
	}

	// Rebuild feedData with all collected posts
	feedData := map[string]interface{}{
		"feed": allPosts,
	}
	if cursor != "" {
		feedData["cursor"] = cursor
	}

	// Format results as markdown
	markdown := t.formatFeedMarkdown(feedData)

	return &mcp.ToolResult{
		Content: []mcp.ContentItem{
			{
				Type: "text",
				Text: markdown,
			},
		},
	}, nil
}

// resolveFeedURI resolves a feed name/query to a full at:// URI
func (t *FeedTool) resolveFeedURI(ctx context.Context, feed string) (string, error) {
	// Check if it's already a valid at:// URI
	if strings.HasPrefix(feed, "at://") && strings.Contains(feed, "/app.bsky.feed.generator/") {
		return feed, nil
	}

	// Not a full URI - search for feed by name
	params := map[string]string{
		"query": feed,
	}

	searchResult, err := t.apiClient.GetPublic(ctx, "app.bsky.unspecced.getPopularFeedGenerators", params)
	if err != nil {
		return "", fmt.Errorf("failed to search for feed: %w", err)
	}

	// Extract feeds from search result
	feeds, ok := searchResult["feeds"].([]interface{})
	if !ok || len(feeds) == 0 {
		return "", fmt.Errorf("no feeds found matching '%s'. Please provide a valid feed URI (at://...) or search term", feed)
	}

	// Get the first feed's URI
	firstFeed, ok := feeds[0].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid feed data in search result")
	}

	uri, ok := firstFeed["uri"].(string)
	if !ok || uri == "" {
		return "", fmt.Errorf("feed URI not found in search result")
	}

	return uri, nil
}

// formatFeedMarkdown formats feed data as markdown per docs/16-mcp-schemas.md spec
func (t *FeedTool) formatFeedMarkdown(feedData map[string]interface{}) string {
	var sb strings.Builder

	// Extract posts from feed
	feedArray, ok := feedData["feed"].([]interface{})
	if !ok || len(feedArray) == 0 {
		sb.WriteString("# Feed · 0 posts\n\nNo posts found in feed.\n")
		return sb.String()
	}

	sb.WriteString(fmt.Sprintf("# Feed · %d posts\n\n", len(feedArray)))

	// Track seen posts for ID compaction
	seenPosts := make(map[string]bool)

	// Format each post
	for _, item := range feedArray {
		feedItem, ok := item.(map[string]interface{})
		if !ok {
			continue
		}

		post, ok := feedItem["post"].(map[string]interface{})
		if !ok {
			continue
		}

		// Extract post data
		uri, _ := post["uri"].(string)
		rkey := ExtractRkey(uri)

		author, _ := post["author"].(map[string]interface{})
		handle, _ := author["handle"].(string)

		record, _ := post["record"].(map[string]interface{})
		text, _ := record["text"].(string)
		createdAt, _ := record["createdAt"].(string)

		likes := GetIntField(post, "likeCount")
		replies := GetIntField(post, "replyCount")
		reposts := GetIntField(post, "repostCount")
		quotes := GetIntField(post, "quoteCount")

		// Author ID line
		fullID := fmt.Sprintf("%s/%s", handle, rkey)
		authorID := CompactPostID(handle, rkey, seenPosts)
		sb.WriteString(fmt.Sprintf("%s\n", authorID))
		seenPosts[fullID] = true

		// Blockquote content
		sb.WriteString(BlockquoteContent(text))
		sb.WriteString("\n")

		// Stats and timestamp
		stats := FormatStats(likes, reposts, quotes, replies)
		timestamp := FormatTimestamp(createdAt)

		if stats != "" {
			sb.WriteString(fmt.Sprintf("%s  %s\n", stats, timestamp))
		} else {
			sb.WriteString(fmt.Sprintf("%s\n", timestamp))
		}

		sb.WriteString("\n")
	}

	// Add cursor information for pagination
	if cursor, ok := feedData["cursor"].(string); ok && cursor != "" {
		sb.WriteString(fmt.Sprintf("**Next cursor:** `%s`\n", cursor))
	}

	return sb.String()
}

// atURIToBskyURL converts an AT URI to a Bluesky web URL
func (t *FeedTool) atURIToBskyURL(atURI string) string {
	// Parse AT URI: at://{did}/{collection}/{rkey}
	if !strings.HasPrefix(atURI, "at://") {
		return atURI
	}

	parts := strings.Split(strings.TrimPrefix(atURI, "at://"), "/")
	if len(parts) < 3 {
		return atURI
	}

	did := parts[0]
	rkey := parts[2]

	return fmt.Sprintf("https://bsky.app/profile/%s/post/%s", did, rkey)
}

// Helper functions for extracting values from maps

func getStringParam(args map[string]interface{}, key, defaultValue string) string {
	if val, ok := args[key]; ok {
		if str, ok := val.(string); ok {
			return strings.TrimSpace(str)
		}
	}
	return defaultValue
}

func getIntParam(args map[string]interface{}, key string, defaultValue int) int {
	if val, ok := args[key]; ok {
		switch v := val.(type) {
		case int:
			return v
		case int64:
			return int(v)
		case float64:
			return int(v)
		}
	}
	return defaultValue
}

func getStringFromMap(m map[string]interface{}, key, defaultValue string) string {
	if val, ok := m[key]; ok {
		if str, ok := val.(string); ok {
			return str
		}
	}
	return defaultValue
}

func getIntFromMap(m map[string]interface{}, key string, defaultValue int) int {
	if val, ok := m[key]; ok {
		switch v := val.(type) {
		case int:
			return v
		case int64:
			return int(v)
		case float64:
			return int(v)
		}
	}
	return defaultValue
}

File: rust-server/src/tools/feed.rs
//! Feed tool implementation
//!
//! Implements the `feed` MCP tool for fetching BlueSky feeds

use crate::cli::FeedArgs;
use crate::error::AppError;
use crate::http::client_with_timeout;
use crate::mcp::{McpResponse, ToolResult};
use anyhow::Result;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::time::Duration;
use tokio::time::timeout;
use tracing::debug;

#[derive(Deserialize, Serialize, Debug, Clone)]
struct PostAuthor {
    did: String,
    handle: String,
    #[serde(rename = "displayName")]
    display_name: Option<String>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
struct PostRecord {
    text: String,
    #[serde(rename = "createdAt")]
    created_at: String,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
struct FeedPost {
    uri: String,
    cid: String,
    author: PostAuthor,
    record: PostRecord,
    #[serde(rename = "indexedAt")]
    indexed_at: String,
    #[serde(rename = "likeCount")]
    like_count: Option<i32>,
    #[serde(rename = "replyCount")]
    reply_count: Option<i32>,
    #[serde(rename = "repostCount")]
    repost_count: Option<i32>,
    #[serde(rename = "quoteCount")]
    quote_count: Option<i32>,
}

#[derive(Deserialize)]
struct FeedViewPost {
    post: FeedPost,
}

#[derive(Deserialize)]
struct FeedResponse {
    feed: Vec<FeedViewPost>,
    cursor: Option<String>,
}

#[derive(Deserialize)]
struct FeedGeneratorView {
    uri: String,
    #[serde(rename = "displayName")]
    display_name: Option<String>,
}

#[derive(Deserialize)]
struct PopularFeedsResponse {
    feeds: Vec<FeedGeneratorView>,
}

/// Resolve a feed name/query to a full at:// URI by searching
async fn resolve_feed_uri(client: &reqwest::Client, query: &str) -> Result<String, AppError> {
    let search_url = format!(
        "https://public.api.bsky.app/xrpc/app.bsky.unspecced.getPopularFeedGenerators?query={}",
        urlencoding::encode(query)
    );

    debug!("Searching for feed with query: {}", query);

    let response = client
        .get(&search_url)
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Failed to search for feed: {}", e)))?;

    if !response.status().is_success() {
        return Err(AppError::NetworkError(format!(
            "Feed search API returned error {}",
            response.status()
        )));
    }

    let search_response: PopularFeedsResponse = response
        .json()
        .await
        .map_err(|e| AppError::ParseError(format!("Failed to parse feed search response: {}", e)))?;

    if search_response.feeds.is_empty() {
        return Err(AppError::InvalidInput(format!(
            "No feeds found matching '{}'. Please provide a valid feed URI (at://...) or search term.",
            query
        )));
    }

    let first_feed = &search_response.feeds[0];
    debug!(
        "Found feed: {} ({})",
        first_feed.display_name.as_deref().unwrap_or("unnamed"),
        first_feed.uri
    );

    Ok(first_feed.uri.clone())
}

/// Handle feed tool call
pub async fn handle_feed(id: Option<Value>, args: Value) -> McpResponse {
    match timeout(Duration::from_secs(120), handle_feed_impl(args)).await {
        Ok(result) => match result {
            Ok(content) => McpResponse::success(id, serde_json::to_value(content).unwrap()),
            Err(e) => McpResponse::error(id, e.error_code(), &e.message()),
        },
        Err(_) => McpResponse::error(id, "timeout", "Feed request exceeded 120 second timeout"),
    }
}

async fn handle_feed_impl(args: Value) -> Result<ToolResult, AppError> {
    let feed_args: FeedArgs = serde_json::from_value(args)
        .map_err(|e| AppError::InvalidInput(format!("Invalid arguments: {}", e)))?;

    execute_feed(feed_args).await
}

/// Execute feed tool
pub async fn execute_feed(feed_args: FeedArgs) -> Result<ToolResult, AppError> {
    debug!("Feed request for feed: {:?}", feed_args.feed);

    let client = client_with_timeout(Duration::from_secs(120));
    
    // Resolve the feed URI
    let feed_uri = match &feed_args.feed {
        Some(feed_input) => {
            // Check if it's already a valid at:// URI
            if feed_input.starts_with("at://") && feed_input.contains("/app.bsky.feed.generator/") {
                feed_input.clone()
            } else {
                // Not a full URI - search for feed by name
                debug!("Feed '{}' is not a full URI, searching...", feed_input);
                resolve_feed_uri(&client, feed_input).await?
            }
        }
        None => {
            // Default to "What's Hot" feed
            "at://did:plc:z72i7hdynmk6r22z27h6tvur/app.bsky.feed.generator/whats-hot".to_string()
        }
    };

    debug!("Using feed URI: {}", feed_uri);

    // Build the URL for the getFeed endpoint
    let mut url = format!(
        "https://public.api.bsky.app/xrpc/app.bsky.feed.getFeed?feed={}",
        urlencoding::encode(&feed_uri)
    );

    if let Some(cursor) = &feed_args.cursor {
        url.push_str(&format!("&cursor={}", urlencoding::encode(cursor)));
    }

    if let Some(limit) = feed_args.limit {
        let clamped_limit = limit.clamp(1, 100);
        url.push_str(&format!("&limit={}", clamped_limit));
    }

    debug!("Fetching feed from: {}", url);

    let response = client
        .get(&url)
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Failed to fetch feed: {}", e)))?;

    if !response.status().is_success() {
        let status = response.status();
        let error_text = response.text().await.unwrap_or_default();
        return Err(AppError::NetworkError(format!(
            "Feed API returned error {}: {}",
            status, error_text
        )));
    }

    let feed_response: FeedResponse = response
        .json()
        .await
        .map_err(|e| AppError::ParseError(format!("Failed to parse feed response: {}", e)))?;

    debug!("Received {} posts from feed", feed_response.feed.len());

    // Format as markdown per docs/16-mcp-schemas.md spec
    let mut markdown = String::new();
    markdown.push_str(&format!("# Feed · {} posts\n\n", feed_response.feed.len()));

    use crate::tools::post_format::*;
    use std::collections::HashMap;
    let mut seen_posts: HashMap<String, String> = HashMap::new();

    for feed_post in &feed_response.feed {
        let post = &feed_post.post;
        let rkey = extract_rkey(&post.uri);
        let full_id = format!("{}/{}", post.author.handle, rkey);
        
        // Author ID line
        let author_id = compact_post_id(&post.author.handle, rkey, &seen_posts);
        markdown.push_str(&format!("{}\n", author_id));
        seen_posts.insert(full_id, post.uri.clone());
        
        // Blockquote content
        markdown.push_str(&blockquote_content(&post.record.text));
        markdown.push('\n');
        
        // Stats and timestamp
        let stats = format_stats(
            post.like_count.unwrap_or(0),
            post.repost_count.unwrap_or(0),
            post.quote_count.unwrap_or(0),
            post.reply_count.unwrap_or(0),
        );
        let timestamp = format_timestamp(&post.record.created_at);
        
        if !stats.is_empty() {
            markdown.push_str(&format!("{}  {}\n", stats, timestamp));
        } else {
            markdown.push_str(&format!("{}\n", timestamp));
        }
        
        markdown.push('\n');
    }

    if let Some(cursor) = feed_response.cursor {
        markdown.push_str(&format!("**Next cursor:** `{}`\n", cursor));
    }

    Ok(ToolResult::text(markdown))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_feed_args_deserialize() {
        let json = serde_json::json!({
            "feed": "at://did:plc:example/app.bsky.feed.generator/test",
            "limit": 50
        });

        let args: FeedArgs = serde_json::from_value(json).unwrap();
        assert_eq!(
            args.feed,
            Some("at://did:plc:example/app.bsky.feed.generator/test".to_string())
        );
        assert_eq!(args.limit, Some(50));
    }

    #[test]
    fn test_feed_args_optional_fields() {
        let json = serde_json::json!({});
        let args: FeedArgs = serde_json::from_value(json).unwrap();
        assert_eq!(args.feed, None);
        assert_eq!(args.limit, None);
        assert_eq!(args.cursor, None);
    }
}
